src/Hopper/Internal/LoweredCore/ANF.hs

```diff
@@ -0,0 +1,477 @@
+{-# LANGUAGE TemplateHaskell #-}
+
+----------------------------------------------------------------------------
+-- |
+-- This module defines an Administrative Normal Form (ANF) representation of a
+-- Hopper abstract syntax tree, 'Anf'. In ANF, only simple allocations or
+-- applications can occur on the right-hand side of a 'Let'. This is similar to
+-- an SSA or CPS form.
+--
+-- The function 'toAnf' lowers a 'Term' AST to 'Anf'. This is achieved by an
+-- adaptation of Olivier Danvy's algorithm from "A New One-Pass Transformation
+-- into Monadic Normal Form" (2002) to our 2D De Bruijn 'Variable'
+-- representation.
+----------------------------------------------------------------------------
+
+module Hopper.Internal.LoweredCore.ANF
+  (
+  -- * Data Types
+    Arity(..)
+  , Anf(..)
+  , App(..)
+  , Alloc(..)
+  , Rhs(..)
+
+  -- * Lowering to ANF
+  , toAnf
+  ) where
+
+import Hopper.Utils.LocallyNameless
+import Hopper.Internal.Core.Literal
+import Hopper.Internal.Core.Term
+
+import Data.Foldable (foldl')
+import Data.Maybe (fromMaybe, isJust)
+import Data.Monoid (Sum(..), Last(..))
+import Data.Word
+import Control.Arrow ((&&&), (***))
+import Control.Monad (replicateM, join)
+import Control.Monad.Trans.Reader (ReaderT, runReaderT, ask, local, reader)
+import Control.Monad.Trans.State.Strict (State, evalState)
+import Control.Lens hiding (levels)
+
+import qualified Data.Vector as V
+import qualified Data.Map.Strict as Map
+
+--------------------------
+-- Data Types
+--------------------------
+
+-- TODO: possibly allow more "atomic" expression types (besides variables) once
+--       we have some unboxed values
+
+-- TODO: switch back away from this
+newtype Arity = Arity Word32 deriving (Eq,Ord,Read,Show)
+
+data Anf
+  = AnfReturn !(V.Vector Variable) -- indices into the current env stack
+  | AnfLet !Arity -- TODO: switch back to !(V.Vector BinderInfo)
+           -- !(Maybe SourcePos)
+           !Rhs
+           !Anf
+  | AnfTailCall !App
+  deriving (Eq,Ord,Read,Show)
+
+data App
+  = AppFun !Variable
+           !(V.Vector Variable)
+  | AppPrim !PrimOpId !(V.Vector Variable)
+  | AppThunk !Variable
+  deriving (Eq,Ord,Read,Show)
+
+data Alloc
+  = AllocLit !Literal
+  | AllocLam !Arity -- TODO: switch back to !(V.Vector BinderInfo)
+             !Anf
+  | AllocThunk !Anf
+  deriving (Eq,Ord,Read,Show)
+
+data Rhs
+  = RhsAlloc !Alloc
+  | RhsApp !App
+  deriving (Eq,Ord,Read,Show)
+
+--------------------------
+-- Lowering to ANF
+--------------------------
+
+-- | The first slot in the most recent binder
+v0 :: Variable
+v0 = LocalVar $ LocalNamelessVar 0 $ BinderSlot 0
```

Comment:
[Yesterday](https://github.com/joelburget/daily-typecheckers/blob/master/src/Day9.hs#L173-L174) I learned this trick from [Chargueraud](http://www.chargueraud.org/research/2009/ln/main.pdf) where it's slightly more expressive to write `let (x, xVAr) = genFresh (freeVars t) in abs x (App xVar t)` than `Abs (App (BVar 0) t)`. In other words, dealing with the higher-level representation of named free variables, rather than de Bruijn indices. I think there's extra win to be had for us, since we deal with the extra bookkeeping of 2d indices.

Note that my daily typecheckers code isn't an especially good example of this, since it should use the helper `abs t = Abs (closeOuter x t)` (which I used earlier in this comment).

You don't necessarily have to do this, especially in this review, just wanted to mention it.